std/internal/math/gammafunction.d

```diff
@@ -537,8 +537,10 @@ unittest {
             assert( feqrel(log(fabs(gamma(testpoints[i]))), testpoints[i+1]) > real.mant_dig-5);
         }
     }
-    assert(logGamma(-50.2) == log(fabs(gamma(-50.2))));
-    assert(logGamma(-0.008) == log(fabs(gamma(-0.008))));
+    version(DragonFlyBSD) {} else {  // FIXME
+       assert(logGamma(-50.2) == log(fabs(gamma(-50.2))));
+       assert(logGamma(-0.008) == log(fabs(gamma(-0.008))));
+    }
```

Comment:
This wasn't need for dmd upstream?  If so, it suggests a problem with the ldc or druntime pulls, likely some math function.

Reply:
This has not been resolved yet. Would be nice to get some help debugging this one (Maybe from code owner).

Reply:
Should be pretty easy to start debugging by checking the intermediate results, most likely a loss of precision somewhere.  Try running this code with dmd and ldc and see what results you get for each:
```
void main() {
	dumpNums(-50.2);
	dumpNums(-0.008);
}

void dumpNums(real x) {
import std.stdio, std.math, std.internal.math.gammafunction;
    writefln("logGamma: %g - %A", logGamma(x), logGamma(x)); 
    writefln("gamma: %g - %A, log: %g - %A", gamma(x), gamma(x), log(fabs(gamma(x))), log(fabs(gamma(x)))); 
}
```
With dmd 2.077.1 and ldc 1.6 run on linux/x64 with 80-bit reals, I get the following output, including hex formatting for the full precision:
```
logGamma: -147.586 - -0X9.39620BA62168DB6P+4
gamma: -8.01724e-65 - -0X8.717187D9E6EC13EP-216, log: -147.586 - -0X9.39620BA62168DB6P+4
logGamma: 4.83298 - 0X9.AA7CEB47625307BP-1
gamma: -125.585 - -0XF.B2B9D9745DC2BP+3, log: 4.83298 - 0X9.AA7CEB47625307BP-1
```

Reply:
@joakim-noah I already started down this road.

DragonFlyBSD         | DMD                       | LDC                       | G++                     |
-------------------- | ------------------------- | ------------------------- | ----------------------- |
logGamma (%g)        | -147.586                  | -147.586                  | -147.586                |
logGamma (%A)        | -0X1.272C4174C42D1B6CP+7  | -0X1.272C4174C42D1B6CP+7  | -0X1.272C4174C42D3P+7   |
gamma (%g)           | -8.01724e-65              | -8.01724e-65              | -8.01724e-65            |
gamma (%A)           | -0X1.0E2E30FB3CDD827CP-213| -0X1.0E2E30FB3CDD827CP-213| -0X1.0E2E30FB3CD69P-213 |
log(fabs(gamma))) %g | -147.586                  | -147.586                  | -147.586                |
log(fabs(gamma))) %A | -0X1.272C4174C42D1B6CP+7  | -0X1.272C4174C42D1B6AP+7  | -0X1.272C4174C42D3P+7   |

Notice the rouding difference between dmd/ldc on the last line.
Notice that the c++ implementation generates slightly different results (%A)

Linux (x86_86)   | DMD                       | LDC                       | G++                     |
-------------------- | ------------------------- | ------------------------- | ----------------------- |
logGamma (%g)        | -147.586                  | -147.586                  | -147.586                |
logGamma (%A)        | -0X9.39620BA62168DB6P+4   | -0X9.39620BA62168DB6P+4   | -0X1.272C4174C42D3P+7   |  
gamma (%g)           | -8.01724e-65              | -8.01724e-65              | -8.01724e-65            |
gamma (%A)           | -0X8.717187D9E6EC13EP-216 | -0X8.717187D9E6EC13EP-216 | -0X1.0E2E30FB3CD69P-213 | 
log(fabs(gamma))) %g | -147.586                  | -147.586                  | -147.586                |
log(fabs(gamma))) %A | -0X9.39620BA62168DB6P+4   | -0X9.39620BA62168DB6P+4   | -0X1.272C4174C42D3P+7   |

Notice that c++ delivers the exact same results on Linux and DragonFly. On Linux, the DMD and LDC results do match up.

C++ program used:
```
#include <iostream>
#include <sstream>
#include <stdio.h>
#include <math.h>

int main(int argc, char **argv)
{
    std::stringstream conversionStream(argv[1]);
    double input;
    conversionStream >> input;
    printf("gamma(%f) =%g  -  %A\n", input, gamma(input), gamma(input));
    printf("lgamma(%f) =%g  -  %A\n", input, lgamma(input), lgamma(input));
    printf("tgamma(%f) =%g  -  %A, log: %g  -  %A\n", input, tgamma(input), tgamma(input), log(fabs(tgamma(input))), log(fabs(tgamma(input))));
    return 0;
}
```
We should be using lgamma/lgammal and tgamma/tgammal instead of gamma/gammal, according to ISO Standard [N1570](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf). Also check [stackoverflow](https://stackoverflow.com/questions/18116376/what-is-the-definition-for-gammadouble-x-and-why-is-it-different-on-two-gcc-ve)

Reply:
OK, if the issue is just some flutter in the last bits because of different log implementations, don't worry about it and take this patch out.  It really doesn't matter if this test doesn't pass on DragonFly, simply because it expects exact equality.

Reply:
- Can you explain why running the same functions using g++ / clang++ produces identical output on linux and dragonfly ? But results using dmd and/or ldc differ between the two. Is this something that would be worth investigating ?
- How come the llvm_instrinsic log output differs from using platform math.h function ?
- But that would mean we "always" fail the unittest (when CI is inplace at some point). I don't think that is ideal. If necessary we can rewrite the unittest to check with less precision.

Reply:
Your second question answers your first: ldc and dmd are likely using different log functions on DragonFly.  As for CI, we can worry about that when it happens, various tests have been made to ignore the last couple bits in `std.math` already.

Reply:
The difference probably comes from DMD using an approximation for `std.math.log()` (based on base-2-log via x87 instruction `yl2x`, independent from OS) and LDC using the LLVM intrinsic mapping to C function `logl()`, which may diverge across OS's. I'd expect consistent results (and potentially better performance) when linking in LLVM's builtins library from compiler-rt, which is supposed to provide tuned implementations of some LLVM intrinsics.

Reply:
@joakim-noah Thanks for your input. I opted to implement a specific version that takes care of the rounding difference, this way the reason for the different handling is clear a preserved for future reference.